scripts/sign_windows_executable.js

```diff
@@ -0,0 +1,169 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+const minimist = require('minimist');
+const fsPromises = require('fs/promises');
+const fs = require('fs');
+const path = require('path');
+const util = require('util');
+
+function assert(condition, msg) {
+  if (!condition) {
+    throw new Error(msg);
+  }
+}
+
+async function assertFileExists(file, msg) {
+  try {
+    await fsPromises.access(file, fs.constants.R_OK | fs.constants.W_OK);
+  } catch (err) {
+    throw new Error(util.format(msg, file), { cause: err });
+  }
+}
+
+/**
+ * Get the required option value from either cliValue or environment variable.
+ * @param {object} options the CLI options object.
+ * @param {string} argName the CLI argument name.
+ * @param {string} envName the environment variable name for this option.
+ * @param {boolean} required indicates whether this option is required.
+ * @returns {string} the value of the option
+ */
+function getOptionValue(options, argName, envName, required) {
+  const cliValue = options ? options[argName] : null;
+  const v = cliValue ?? process.env[envName];
+  if (required) {
+    assert(!!v, `either --${argName} or ${envName} is required`);
+  }
+  return v;
+}
+
+function appendPfxJsignArgs(args, options) {
+  // self-signed development certificate
+  args.push('--storetype', 'PKCS12');
+
+  const pfxCert = getOptionValue(options, 'pfx',
+    'WINDOWS_SIGNING_PFX_CERT', true);
+  args.push('--keystore', pfxCert);
+}
+
+function appendDigicertUsbJsignArgs(args, options) {
+  // extended validation certificate stored in USB drive
+  args.push('--storetype', 'PKCS11');
+
+  const subject = getOptionValue(options, 'subject',
+    'WINDOWS_SIGNING_EV_CERT_SUBJECT', false);
+  if (subject) {
+    args.push('--alias', subject);
+  }
+
+  switch (process.platform) {
+    case 'win32':
+      args.push('--keystore', 'scripts/digicert-usb-config/eToken-windows.cfg');
+      break;
+    case 'darwin':
+      args.push('--keystore', 'scripts/digicert-usb-config/eToken-macos.cfg');
+      break;
+    default:
+      throw new Error(`we do not support ev signing on ${process.platform}`);
+  }
+}
+
+/**
+ * Sign the target exeFile using a specific algorithm and options.
+ * @param {string} exeFile the full path of the exe file to be signed.
+ * @param {'sha1'|'sha256'} algorithm the algorithm used for signing.
+ * @param {object} options the additional options (cli arguments) for signing.
+ */
+async function signWindowsExecutable(exeFile, algorithm, options) {
+  const type = getOptionValue(options, 'certtype',
+    'WINDOWS_SIGNING_CERT_TYPE', false);
+  if (!type || type == 'none') {
+    console.info(`skip signing "${exeFile}"`);
+    return;
+  }
+
+  assert(!!exeFile, 'executable path is required');
```

Comment:
Even better!

Reply:
Just make sure to use `??` instead of `||` in cases where `undefined` or `null` are the only things you want to check for.

Reply:
For instance, an imaginary bug where `""` is implied to be the root folder, this would error w/o `??` (a terrible api choice, though)

Reply:
Actually `throw` is not an expression currently in JS/TS (but there is a [proposal](https://github.com/tc39/proposal-throw-expressions)), therefore this expression triggers syntax error: Expression Expected. We can have workaround to wrap it into a function as mentioned [here](https://stackoverflow.com/questions/60884426/how-to-check-for-null-value-inline-and-throw-an-error-in-typescript). Otherwise we have to use a tedious `if` statement.